url696d9a93c9_liaosiwei_bayesnetwork_tgz-pJ8-tutorial_TutorialJ8/src/com/alibaba/Train.java

```diff
@@ -0,0 +1,352 @@
+        c.setParents(Arrays.asList(a));
+
+
+        d = net.createNode("d");
+        d.addOutcomes("true", "false");
+        d.setParents(Arrays.asList(a, b, c));
+
+        inferer.getFactory().setUseLogScale(true);
+        inferer.getFactory().setFloatingPointType(float.class);
+    }
+
+    public static void initNetwork() {
+        a.setProbabilities(clickProb);
+        b.setProbabilities(saveProb);
+        c.setProbabilities(carProb);
+        d.setProbabilities(buyProb);
+
+
+        inferer.setNetwork(net);
+    }
+
+    public static void setEvidence(Set<Integer> s) {
+        Map<BayesNode,String> evidence = new HashMap<BayesNode,String>();
+        for (Integer i: s) {
+            if (i == 0) evidence.put(a, "true");
+            else if (i == 2) evidence.put(b, "true");
+            else if (i == 3) evidence.put(c, "true");
+        }
+        inferer.setEvidence(evidence);
+    }
+
+    public static Boolean isExpendable() {
+        try {
+            double [] belifes = inferer.getBeliefs(d);
+/*            System.out.println(belifes[0] + " " + belifes[1]);*/
+            return belifes[0] > belifes[1];
+        }
+        catch (NumericalInstabilityException e) {
+/*            System.out.println(e.getMessage());*/
+            return false;
+        }
+    }
+
+    public static void main(String[] args) {
+        DataRead reader = new DataRead();
+        createNetwork();
+        if (!reader.read()) {
+            System.out.println("read file error");
+            return;
+        }
+        Map<String, Map<String, Map<Integer, HashSet<Integer>>>> table = reader.getData();
+
+        long totalSum = 0;
+        for (String key1: table.keySet()) {
+            Set<String> allSet = new HashSet<String>();
+            Set<String> clickSet = new HashSet<String>();
+            Set<String> carSet = new HashSet<String>();
+            Set<String> saveSet = new HashSet<String>();
+            Set<String> buySet = new HashSet<String>();
+
+            /*
+            * date format:
+            *
+            * click = true
+            *   save = true
+            *     car = true
+            *     0.2, 0.8 // buy=true and buy=false
+            *     car = false
+            *     0.5, 0.5 // true and false
+            *   save = false
+            *     car = true
+            *     X, X // true and false
+            *     car = false
+            *     X, X // true and false
+            *
+            * click = false
+            *   save = true
+            *     car = true
+            *     0.2, 0.8 // buy=true and buy=false
+            *     car = false
+            *     0.5, 0.5 // true and false
+            *   save = false
+            *     car = true
+            *     X, X // true and false
+            *     car = false
+            *     X, X // true and false
+            *
+            *
+            * */
+
+            for (String key2: table.get(key1).keySet()) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.